pkg/verification/verification_service.go

```diff
@@ -147,56 +189,78 @@ func generateVerificationCode() (string, error) {
+		if parseErr != nil {
+			// If the verification expiry timestamp is corrupt or missing, then return an error
+			verificationErr = errors.NewInternalError(parseErr, "error parsing expiry timestamp")
+		} else if now.After(exp) {
+			// If it is now past the expiry timestamp for the verification code, return a 403 Forbidden error
+			verificationErr = errors.NewForbiddenError("expired", "verification code expired")
+		}
 	}
 
-	if now.After(exp) {
-		// If it is now past the expiry timestamp for the verification code, return a 403 Forbidden error
-		return signup, errors.NewForbiddenError("expired", "verification code expired")
+	if verificationErr == nil {
+		if code != signup.Annotations[v1alpha1.UserSignupVerificationCodeAnnotationKey] {
+			// The code doesn't match
+			attemptsMade++
+			signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = strconv.Itoa(attemptsMade)
+			verificationErr = errors.NewForbiddenError("invalid code", "the provided code is invalid")
+		}
 	}
 
-	if code != signup.Annotations[v1alpha1.UserSignupVerificationCodeAnnotationKey] {
-		// The code doesn't match
-		attemptsMade++
-		signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = strconv.Itoa(attemptsMade)
-		return signup, errors.NewForbiddenError("invalid code", "the provided code is invalid")
+	if verificationErr == nil {
+		// If the code matches then set VerificationRequired to false, reset other verification annotations
+		signup.Spec.VerificationRequired = false
+		delete(signup.Annotations, v1alpha1.UserSignupVerificationCodeAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserVerificationAttemptsAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserSignupVerificationCounterAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserSignupVerificationTimestampAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserSignupVerificationInitTimestampAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserVerificationExpiryAnnotationKey)
+	} else {
+		log.Error(ctx, verificationErr, "error validating verification code")
 	}
 
-	// If the code matches then set VerificationRequired to false, reset other verification annotations
-	signup.Spec.VerificationRequired = false
-	delete(signup.Annotations, v1alpha1.UserSignupVerificationCodeAnnotationKey)
-	delete(signup.Annotations, v1alpha1.UserVerificationAttemptsAnnotationKey)
-	delete(signup.Annotations, v1alpha1.UserSignupVerificationCounterAnnotationKey)
-	delete(signup.Annotations, v1alpha1.UserSignupVerificationTimestampAnnotationKey)
-	delete(signup.Annotations, v1alpha1.UserSignupVerificationInitTimestampAnnotationKey)
-	delete(signup.Annotations, v1alpha1.UserVerificationExpiryAnnotationKey)
+	// Update changes made to the UserSignup
+	_, updateErr := s.Services().SignupService().UpdateUserSignup(signup)
+	if updateErr != nil {
+		log.Error(ctx, updateErr, "error updating UserSignup")
+		verificationErr = updateErr
+	}
 
-	return signup, nil
+	return
```

Comment:
it looks like this function swallows the errors, it would probably be beneficial to return it?